packages/world/src/systems/MineSystem.sol

```diff
@@ -93,93 +93,66 @@ contract MineSystem is System {
+      minedType = RandomResourceLib._collapseRandomOre(mined, coord);
     }
 
-    // First coord will be the base coord, the rest is relative schema coords
-    Vec3[] memory coords = mineObjectTypeId.getRelativeCoords(baseCoord, Orientation._get(mined));
+    MineLib._requireMinesAllowed(caller, minedType, coord, extraData);
 
     uint128 finalMass = MassReductionLib._processMassReduction(caller, mined, toolSlot);
 
-    if (finalMass == 0) {
-      if (mineObjectTypeId == ObjectTypes.Bed) {
-        // If mining a bed with a sleeping player, kill the player
-        MineLib._mineBed(mined, baseCoord);
-      }
-
-      if (bytes(DisplayURI._get(mined)).length > 0) {
-        DisplayURI._deleteRecord(mined);
-      }
-
-      Mass._deleteRecord(mined);
-
-      {
-        // Remove seeds placed on top of this block
-        Vec3 aboveCoord = baseCoord + vec3(0, 1, 0);
-        (EntityId above, ObjectTypeId aboveTypeId) = getEntityAt(aboveCoord);
-        if (aboveTypeId.isSeed()) {
-          _requireSeedNotFullyGrown(mined);
-          if (!above.exists()) {
-            above = createEntityAt(aboveCoord, aboveTypeId);
-          }
-          _removeBlock(above, aboveCoord);
-          _handleDrop(caller, aboveTypeId, aboveCoord);
-        }
-      }
-
-      _removeBlock(mined, baseCoord);
-      _handleDrop(caller, mineObjectTypeId, baseCoord);
-
-      // If object being mined is a seed, return its energy to local pool
-      if (mineObjectTypeId.isSeed()) {
-        addEnergyToLocalPool(coord, ObjectTypeMetadata._getEnergy(mineObjectTypeId));
-      }
-
-      // Only iterate through relative schema coords
-      for (uint256 i = 1; i < coords.length; i++) {
-        Vec3 relativeCoord = coords[i];
-        (EntityId relative,) = getEntityAt(relativeCoord);
-        BaseEntity._deleteRecord(relative);
-
-        _removeBlock(relative, relativeCoord);
-      }
-    } else {
+    if (finalMass != 0) {
       Mass._setMass(mined, finalMass);
+      return mined;
     }
 
-    MineLib._requireMinesAllowed(caller, mineObjectTypeId, coord, extraData);
-
-    if (finalMass == 0) {
-      // Detach program if it exists
-      ProgramId program = mined.getProgram();
-      if (program.exists()) {
-        bytes memory onDetachProgram = abi.encodeCall(IDetachProgramHook.onDetachProgram, (caller, mined, extraData));
-        program.call({ gas: SAFE_PROGRAM_GAS, hook: onDetachProgram });
-      }
+    // The block was fully mined
+    Mass._deleteRecord(mined);
 
-      if (mineObjectTypeId == ObjectTypes.ForceField) {
-        destroyForceField(mined);
-      }
+    // Remove flora on top of this block
+    Vec3 aboveCoord = baseCoord + vec3(0, 1, 0);
+    // If above is a seed, the entity must exist as there are not seeds in the base terrain
+    (EntityId above, ObjectTypeId aboveTypeId) = getOrCreateEntityAt(aboveCoord);
+    if (aboveTypeId.isFlora()) {
+      _removeFlora(above, aboveTypeId, aboveCoord);
```

Comment:
We're making it so you cant have any floating flowers?

But this doesnt cascade right?

Reply:
yeah this shouldn't cascade.. but now that I think about it we do have passthrough blocks on top of each other in the terrain right? like tall grass maybe? Also, should it be possible to place a passthrough block on top of another one?

Reply:
yea tall grass, cactus, same with vines coming down from trees

Reply:
>  Also, should it be possible to place a passthrough block on top of another one?
imo it should

Reply:
instead of doing this for all non-passthrough blocks, we could just do it for seeds

Reply:
Ah interesting, so @alvrs mentioned that the client had the concept of flora, which I assumed was assigned to all pass through blocks. Are vines also flora in the client? I think it is easy to cascade the remove flora thing but we shouldn't do it for hanging flora

Reply:
The client's concept of "flora" is to render blocks that arent cubic. Not all these blocks are passthrough though. For eg, leaves, bamboo bush, and cactus are not pass through

Vines are indeed flora

If we cascade, could I create a tower of flowers and then mining the bottom flower would revert cuz there's too much gas used?

Reply:
I think only doing it for passthrough (so ie not for cactus) and only doing it for the one right above (or maybe up to depth 2 to catch tall grass) is fine. For vines we actually would not want to cascade all of it, since they're usually hanging down from something. Think it's fine if there are air blocks/flora in some cases, just feels we should handle the default case of wheat/grass/flower on top of a block.

Reply:
could also just say "if the top block was tall grass, remove one more" to avoid always having to check two blocks

Reply:
> But I could place a tall grass and then a solid block above, and the solid block would incorrectly be mined then

i mean only attempt mining the block above that if it is also tall grass haha

> also how do force field permission checks work here? What if the flower is in a different force field from the grass block below? You might have permission to mine the grass block but not the flower above

hm yeah that's almost a game design question, could argue that it should still be mined because it dependend on the block below it, or could argue that it should be skipped because it's protected by the force field. How does this work for beds (or other things that are larger than a single block)? Would tend towards removing it even though it's in a force field.

Reply:
> hm yeah that's almost a game design question

yea deff, I think either could make sense!

For beds (and other multi voxel objects), you can mine any of the individual voxels. If you mine the voxel inside the force field --> force field is called, but if you mine the voxel outside, force field is not called. So its your responsibility as a player to ensure when you place a bed that its fully inside, otherwise it can be mined

Reply:
personally i would go with:
- if block above the mined block is passthrough: destroy it, no matter what
- if the block above the mined block is tall grass, also check the one above that and destroy it if it's also tall grass

alternatively i'd also be fine with:
- only destroy the block above if it's a seed (because in that case it already depends on the farmland)

Reply:
> if the block above the mined block is tall grass, also check the one above that and destroy it if it's also tall grass

what should happen if I have 3 blocks stacked of tall grass? the very top one would remain floating?

Reply:
yeah exactly

Reply:
reverted to just doing it for seeds for now